addExtensionsToHCalAllegro

mmlynari opened the pull request:
BEGINRELEASENOTES
- added new HCalEndcaps_ThreeParts_TileCal_v02.xml: migrated to use FCCSWGridPhiTheta_k4geo; fixed radial dimensions, so the outer radius of all three parts is the same; renamed nModules to nsegments for number of layers in the second cylinder; uses geometry CaloThreePartsEndcap_o1_v02

- added new HCalBarrel_TileCal_v02.xml which uses geometry HCalTileBarrel_o1_v01

- updated ALLEGRO_o1_v03.xml to include HCalBarrel_TileCal_v02.xml and HCalEndcaps_ThreeParts_TileCal_v02.xml

- added new HCalThreePartsEndcap_o1_v02_geo.cpp: added extension to store the radii of each radial layer as well as dimensions of cells. These will be used by the CellPositionsHCalPhiThetaSegTool to calculate the radii of each layer. Improved code readability and variables naming 

- updated HCalTileBarrel_o1_v01_geo.cpp: added extension to store the radii of each radial layer as well as dimensions of cells. These will be used by the CellPositionsHCalPhiThetaSegTool to calculate the radii of each layer. Improved code readability and variables naming

ENDRELEASENOTES

mmlynari commented:
Tagging @faltovaj , @giovannimarchiori and @BrieucF to have a look at the proposed changes

andresailer commented on detector/calorimeter/HCalTileBarrel_o1_v01_geo.cpp:
```diff
@@ -40,7 +41,7 @@ static dd4hep::Ref_t createHCal(dd4hep::Detector& lcdd, xml_det_t xmlDet, dd4hep
   double dSteelSupport = xSteelSupport.thickness();
 
-  lLog << MSG::DEBUG << "steel support thickness: " << dSteelSupport << " [cm]" << endmsg;
+  lLog << MSG::DEBUG << "steel support thickness (cm): " << dSteelSupport << endmsg;
```
```suggestion
  lLog << MSG::DEBUG << "steel support thickness (cm): " << dSteelSupport / dd4hep::cm << endmsg;
```

andresailer commented on detector/calorimeter/HCalTileBarrel_o1_v01_geo.cpp:
```diff
@@ -50,17 +51,15 @@ static dd4hep::Ref_t createHCal(dd4hep::Detector& lcdd, xml_det_t xmlDet, dd4hep
   Dimension sequenceDimensions(sequences[1].dimensions());
   double dzSequence = sequenceDimensions.dz();
-  lLog << MSG::DEBUG << "sequence thickness " << dzSequence << endmsg;
+  lLog << MSG::DEBUG << "sequence thickness (cm) " << dzSequence << endmsg;
```
```suggestion
  lLog << MSG::DEBUG << "sequence thickness (cm) " << dzSequence / dd4hep::cm << endmsg;
```

mmlynari replied:
thanks for the suggestion, changed

andresailer commented on detector/calorimeter/HCalTileBarrel_o1_v01_geo.cpp:
```diff
@@ -223,12 +224,39 @@ static dd4hep::Ref_t createHCal(dd4hep::Detector& lcdd, xml_det_t xmlDet, dd4hep
+  for (unsigned int idxLayer = 0; idxLayer < layerDepths.size(); ++idxLayer) {
+        const double difference_bet_r1r2 = layerDepths.at(idxLayer); 
+
+        caloLayer.distance                  = sensitiveBarrelRmin; // should this be always the radius of the first layer as it is now?  
```
No, this should be the distance to the current layer

mmlynari replied:
Thanks for a clarification, will be corrected in the next commit

andresailer commented on detector/calorimeter/HCalTileBarrel_o1_v01_geo.cpp:
```diff
@@ -223,12 +224,39 @@ static dd4hep::Ref_t createHCal(dd4hep::Detector& lcdd, xml_det_t xmlDet, dd4hep
+        const double difference_bet_r1r2 = layerDepths.at(idxLayer); 
+
+        caloLayer.distance                  = sensitiveBarrelRmin; // should this be always the radius of the first layer as it is now?  
+        caloLayer.sensitive_thickness       = difference_bet_r1r2; // not really sure what is this variable 
```
The thickness of the sensitive element, more or less in radial direction.
This is trivial for sandwich calorimeters where you have a silicon sensor for example.
I don't know what your calorimeter looks like if this is also trivial to figure out.

BrieucF commented on FCCee/ALLEGRO/compact/ALLEGRO_o1_v03/HCalEndcaps_ThreeParts_TileCal.xml:
```diff
@@ -44,20 +44,23 @@
+   <readout name="HCalEndcapReadout_phitheta">
+     <segmentation type="FCCSWGridPhiTheta_k4geo" grid_size_theta="0.022180" phi_bins="EndcapHCal_n_phi_modules" offset_theta="0.06" offset_phi="-pi+(pi/EndcapHCal_n_phi_modules)"/>
+     <id>system:4,type:2,layer:5,theta:11,phi:10</id>
     </readout>
    <readout name="HCalEndcapReadout_phi">
-     <!-- the eta segmentation here is a dummy - the whole hcal is contained in one bin. the z position is stored in the "row" field. -->
-     <segmentation type="FCCSWGridPhiEta_k4geo" grid_size_eta="4." phi_bins="EndcapHCal_n_phi_modules" offset_eta="-3.5" offset_phi="-pi+(pi/EndcapHCal_n_phi_modules)"/>
-     <id>system:4,type:2,layer:5,row:9,eta:10,phi:10</id>
+     <segmentation type="FCCSWGridPhiTheta_k4geo" grid_size_theta="3." phi_bins="EndcapHCal_n_phi_modules" offset_theta="0.06" offset_phi="-pi+(pi/EndcapHCal_n_phi_modules)"/>
+     <id>system:4,type:2,layer:5,row:9,theta:0,phi:10</id>
```
Are those two readouts still needed?

mmlynari replied:
They are not used for the moment, but I'd prefer to keep them to be on the safe side in case they'd be useful in the future. For the HCal barrel, we also keep three readouts.

giovannimarchiori replied:
I think the _phi readout is not needed, so we might get rid of it if cleanup is needed. phi-theta might be needed for clustering, so I'd rather keep if for the time being till we sort out exactly which cells we use for the clustering (do we use cells based on the physical row-phi arrangement or based on a projective theta-phi grid?)

mmlynari replied:
_phi readout removed from both barrel and endcap xml

BrieucF commented on detector/calorimeter/HCalThreePartsEndcap_o1_v01_geo.cpp:
Do you know why the diff of this file is so big?

mmlynari replied:
Originally the file included two functions, buildEC and createHCalEC. To include the layer extensions, several variables from buildEC and also from createHCalEC were needed, so in the new version, we have only one function createHCalEC in which the info from buildEC and createHCalEC is merged.

giovannimarchiori commented on detector/calorimeter/HCalTileBarrel_o1_v01_geo.cpp:
```diff
@@ -223,12 +223,39 @@ static dd4hep::Ref_t createHCal(dd4hep::Detector& lcdd, xml_det_t xmlDet, dd4hep
-  hCal.setPlacement(placedHCal);
-  return hCal;
+  PlacedVolume envelopePhysVol = motherVol.placeVolume(envelopeVolume);
+  envelopePhysVol.addPhysVolID("system", xmlDet.id());
```
just to be sure, since you modified this: do xmlDet.id() and DetElement caloDetElem(xmlDet.nameStr(), xmlDet.id()).id() return the same thing?

mmlynari replied:
they do return the same thing, but just to be on the safe side I replaced xmlDet.id() by caloDetElem.id()

giovannimarchiori commented on detector/calorimeter/HCalThreePartsEndcap_o1_v01_geo.cpp:
```diff
@@ -1,467 +1,496 @@
+    // Make DetElement
+    dd4hep::DetElement caloDetElem(detName, xmlDet.id());
+
+    // Make volume that envelopes the whole barrel; set material to air
```
barrel -> endcap

mmlynari replied:
fixed

giovannimarchiori commented on detector/calorimeter/HCalThreePartsEndcap_o1_v01_geo.cpp:
```diff
@@ -1,467 +1,496 @@
+    Dimension sensDetType = xmlElement.child(_Unicode(sensitive));
+    sensDet.setType(sensDetType.typeStr());
+
+    // Dimension dimensions(xmlElement.child(_Unicode(dimensions)));
```
remove commented line
possibly add a comment to explain meaning of retrieved dimensions in following lines

mmlynari replied:
done

giovannimarchiori commented on detector/calorimeter/HCalThreePartsEndcap_o1_v01_geo.cpp:
```diff
@@ -1,467 +1,496 @@
+    double extBarrelOffset3 = dimensions.v_offset();
+
+    // Hard-coded assumption that we have two different sequences for the modules
+    std::vector<xml_comp_t> sequences = {xmlElement.child(_Unicode(sequence_a)), xmlElement.child(_Unicode(sequence_b))};
```
is it possible to make the code fail if there aren't two sequences?

mmlynari replied:
yes, done

giovannimarchiori commented on detector/calorimeter/HCalThreePartsEndcap_o1_v01_geo.cpp:
```diff
@@ -1,467 +1,496 @@
+
+    // Hard-coded assumption that we have two different sequences for the modules
+    std::vector<xml_comp_t> sequences = {xmlElement.child(_Unicode(sequence_a)), xmlElement.child(_Unicode(sequence_b))};
+    // NOTE: This assumes that both have the same dimensions!
```
is it possible to make the code fail if sequences[0] and [1] have different dimensions?

giovannimarchiori commented on detector/calorimeter/HCalThreePartsEndcap_o1_v01_geo.cpp:
```diff
@@ -1,467 +1,496 @@
+        xml_comp_t layer = *it;
+        Dimension layerDimension(layer.dimensions());
+        numSequencesR1 += layerDimension.nmodules();
+        numSequencesR2 += layerDimension.nModules();
```
I must say that having two functions (and two xml tags) whose name differs only by the capitalisation of a letter is not a very good idea... have you considered having more expressive names?

mmlynari replied:
Agreed, I replaced nModules by nsegments as in the current implementation, the name needs to be a member of struct dd4hep::xml::Dimension. So it's not really easy to find something more meaningful.

andresailer replied:
> the name needs to be a member of struct dd4hep::xml::Dimension

Hmm,

```
layerDimension.child(_U(whatever))
```

should work?

mmlynari replied:
thanks @andresailer , I'd prefer to keep it is as it for now and will try to implement this in the next version of the code

andresailer replied:
Apologies for giving incorrect suggestions!
```c++
    const double whatever    = dimensions.attr< int >("whatever");
```
This actually compiles

giovannimarchiori commented on detector/calorimeter/HCalTileBarrel_o1_v01_geo.cpp:
```diff
@@ -223,12 +223,39 @@ static dd4hep::Ref_t createHCal(dd4hep::Detector& lcdd, xml_det_t xmlDet, dd4hep
+  caloDetElem.addExtension<dd4hep::rec::LayeredCalorimeterData>(caloData);
+
+  caloData->extent[0] = sensitiveBarrelRmin;
+  caloData->extent[1] = sensitiveBarrelRmax; // or r_max ?
```
when you have found out the answer to this question, fix code and remove comment

giovannimarchiori commented on detector/calorimeter/HCalThreePartsEndcap_o1_v01_geo.cpp:
```diff
@@ -1,467 +1,496 @@
+    lLog << MSG::INFO << "dz third part EC:" << dzDetector2 * 2 << endmsg;
+
+
+    for (int iSign = 0; iSign < 2; iSign++){
```
could have simplified this to for (int sign = -1; sign < 2; side+=2) { ..

andresailer commented on detector/calorimeter/HCalTileBarrel_o1_v01_geo.cpp:
```diff
@@ -40,55 +41,66 @@ static dd4hep::Ref_t createHCal(dd4hep::Detector& lcdd, xml_det_t xmlDet, dd4hep
   double sensitiveBarrelRmin = xDimensions.rmin() + xFacePlate.thickness() + space;
 
   // Hard-coded assumption that we have two different sequences for the modules
   std::vector<xml_comp_t> sequences = {xmlDet.child(_Unicode(sequence_a)), xmlDet.child(_Unicode(sequence_b))};
```
You can check if the xmlDet has the childs instead of the check in line 52
```
x_slice.hasAttr(_U(sequence_a)) && x_slice.hasAttr(_U(sequence_b))
```

BrieucF commented:
Can you resolve the conflicts so that we can trigger the tests?

mmlynari commented:
Thanks @BrieucF , the conflicts should be fixed now.

mmlynari commented:
Hello @BrieucF , I updated the barrel xml file to have v02, as the version test was failing due to this. I also rebased my branch and updated the Allegro v03 xml file to include v02 of hcal xmls.

giovannimarchiori commented:
Hello @mmlynari ,

you need to rebase your branch to pickup some recent PRs that introduced some fixes to remove compiler warnings (that are being treated as errors recently and are making the tests fail)

Cheers, Giovanni

BrieucF commented:
Hi Michaela, can you
- remove the old xml files from the folder? `HCalBarrel_TileCal.xml` and `HCalEndcaps_ThreeParts_TileCal.xml`
- update [this](https://github.com/key4hep/k4geo/tree/main/FCCee/ALLEGRO/compact#readme) README with what you changed to ALLEGRO v03
- and [this](https://github.com/key4hep/k4geo/tree/main/detector/calorimeter#readme) README to describe `detector/calorimeter/HCalThreePartsEndcap_o1_v02_geo.cpp`?

mmlynari commented:
Hello @BrieucF , thank you. It's done now

BrieucF commented on detector/calorimeter/README.md:
```diff
@@ -34,13 +34,16 @@ This sub-detector makes calorimeter barrel. It is used in ALLEGRO detector conce
 Original version taken from [FCCDetectors](https://github.com/HEP-FCC/FCCDetectors/blob/70a989a6fc333610e3b1b979c3596da9c41543d8/Detector/DetFCCeeHCalTile/src/HCalThreePartsEndcap_geo.cpp#L4). 
 
+### o1_v02
+Changes wrt o1_v01: Added extension (LayeredCalorimeterData) to store radial layer radii and dimensions. To make this work, the whole code had to be rewritten, but its functionality remains the same. 
```
In the current status of the PR, v01 also has the extension. By reading both comments from v01 and v02, it is not clear what is the difference between them... It feels like they would end-up being exactly the same file (especially "To make this work, the whole code had to be rewritten" which then implies that v01 also had to be completely rewritten since it also has the extension). Do we actually need to modify v01? If yes, do we need v02?

mmlynari replied:
Sorry for the confusion @BrieucF . it should be only HCalThreePartsEndcap_o1_v02 that is changed and includes the extension now, while v01 is untouched and without the extension. Looking back now, what might be confusing is that I added extension to HCalTileBarrel_o1_v01_geo.cpp and probably it would be cleaner to have v02 as the version with extensions for both barrel and endcap. what do you think?

BrieucF replied:
Yes, let's apply the changes in v02 and keep v01 untouched (and update the README accordingly)

mmlynari replied:
Yeah @BrieucF  I agree that it'll be better this way and it's done in the last commit

giovannimarchiori commented:
hi @andresailer,
the PR was approved by @mmlynari. Is there anything else needed (apart from rebasing) for merging?
Thanks a lot!
  Giovanni
